File: delay_action_py/delay_action_py/action_server.py
from action_interfaces.srv import String

import rclpy
from rclpy.node import Node

class ActionServer(Node):
    def __init__(self):
        super().__init__("action_server")
        self.service_ = self.create_service(String, 'action', self.action_delay_cb)
        self.timer = None

        self.get_logger().info("Action Service Ready")

    def action_delay_cb(self, request, response):
        if self.timer is not None:
            self.get_logger().info("Cancelling previous goal")
            self.timer.cancel()
            self.timer = None
        
        self.get_logger().info("Scheduling new future goal. Goal: '%s'" % request.data)
            
        self.timer = self.create_timer(5, lambda : self.timer_cb(request))

        response.output = "Requested goal scheduled. Goal: '%s'" % request.data
        return response

    def timer_cb(self, request):
        self.action_cb(request)

        if self.timer is not None:
            self.timer.cancel()
            self.timer = None

    def action_cb(self, request):
        self.get_logger().info("Goal Activated: '%s'" % request.data)

def main(args=None):
    rclpy.init(args=args)

    action_server = ActionServer()

    rclpy.spin(action_server)
    
    action_server.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: delay_action_cpp/src/action_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "action_interfaces/srv/string.hpp"

#include <chrono>
#include <memory>

using namespace std::chrono_literals;


class ActionServer : public rclcpp::Node
{
  public:
    ActionServer() : Node("action_server")
    {
      service_ = this->create_service<action_interfaces::srv::String>(
        "action",
        std::bind(&ActionServer::action_delay_cb, this, std::placeholders::_1, std::placeholders::_2)
      );

      RCLCPP_INFO(this->get_logger(), "Action Service Ready");
    }

  private:
    void action_delay_cb(const std::shared_ptr<action_interfaces::srv::String::Request> request, std::shared_ptr<action_interfaces::srv::String::Response> response)
    {
      if (timer_) 
      {
        RCLCPP_INFO(this->get_logger(), "Cancelling previous goal");
        timer_->cancel();
      }

      RCLCPP_INFO(this->get_logger(), "Scheduling new future goal. Goal: '%s'", request->data.c_str());
      timer_ = this->create_wall_timer(5s, [this, request]() {
        this->action_cb(request);

        timer_->cancel();
      });

      response->output =  "Requested goal scheduled. Goal: '" + request->data + "'";
    }
    
    void action_cb(const std::shared_ptr<action_interfaces::srv::String::Request> request)
    {    
      RCLCPP_INFO(this->get_logger(), "Goal Activated: '%s'", request->data.c_str());
    }
    
    rclcpp::Service<action_interfaces::srv::String>::SharedPtr service_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  // RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Action Service Ready");
  
  rclcpp::spin(std::make_shared<ActionServer>());

  rclcpp::shutdown();
}